src/mono/mono/mini/interp/jiterpreter.c

```diff
@@ -0,0 +1,1083 @@
+ * To determine whether it is appropriate to insert an entry point at a given candidate location
+ *  we have to scan through all the instructions to estimate whether it is possible to generate
+ *  a suitably large trace. If it's not, we should avoid the overhead of the jiterpreter nop
+ *  instruction that would end up there instead and not waste any resources trying to compile it.
+ */
+void
+jiterp_insert_entry_points (void *_td)
+{
+	if (!jiterpreter_traces_enabled)
+		return;
+	TransformData *td = (TransformData *)_td;
+
+	// Insert an entry opcode for the next basic block (call resume and first bb)
+	// FIXME: Should we do this based on relationships between BBs instead of insn sequence?
+	gboolean enter_at_next = TRUE;
+
+	for (InterpBasicBlock *bb = td->entry_bb; bb != NULL; bb = bb->next_bb) {
+		// Enter trace at top of functions
+		gboolean is_backwards_branch = FALSE,
+			is_resume_or_first = enter_at_next;
+
+		// If backwards branches target a block, enter a trace there so that
+		//  after the backward branch we can re-enter jitted code
+		if (jiterpreter_backward_branch_entries_enabled && bb->backwards_branch_target)
+			is_backwards_branch = TRUE;
+
+		gboolean enabled = (is_backwards_branch || is_resume_or_first);
+		// FIXME: This scan will likely proceed forward all the way out of the current block,
+		//  which means that for large methods we will sometimes scan the same instruction
+		//  multiple times and waste some work. At present this is unavoidable because
+		//  control flow means we can end up with two traces covering different subsets
+		//  of the same method in order to handle loops and resuming
+		gboolean should_generate = enabled && should_generate_trace_here(bb, td->last_ins);
+
+		if (jiterpreter_call_resume_enabled && bb->contains_call_instruction)
+			enter_at_next = TRUE;
+
+		if (jiterpreter_always_generate)
+			should_generate = TRUE;
+
+		if (enabled && should_generate) {
+			td->cbb = bb;
+			interp_insert_ins (td, NULL, MINT_TIER_PREPARE_JITERPRETER);
+			// Note that we only clear enter_at_next here, after generating a trace.
+			// This means that the flag will stay set intentionally if we keep failing
+			//  to generate traces, perhaps due to a string of small basic blocks
+			//  or multiple call instructions.
+			enter_at_next = bb->contains_call_instruction;
+		}
+	}
+}
+
+// Used to parse runtime options that control the jiterpreter. This is *also* used at runtime
+//  by the jiterpreter typescript to reconfigure the jiterpreter, for example if WASM EH is not
+//  actually available even though it was enabled (to turn it off).
+EMSCRIPTEN_KEEPALIVE gboolean
+mono_jiterp_parse_option (const char *option)
+{
+	if (!option || (*option == 0))
+		return FALSE;
+
+	const char *mtl = "jiterpreter-minimum-trace-length=";
+	const int mtl_l = strlen(mtl);
+
+	if (!strcmp (option, "jiterpreter-enable-traces"))
+		jiterpreter_traces_enabled = TRUE;
```

Comment:
It would be better to add these as options to utils/options-def.h. It would auto generate the argument parsing code, etc.